File: windows_app/src/crypto/doubleratchet.cpp
#include "doubleratchet.h"
#include "encryption.h"
#include <QJsonDocument>
#include <QJsonObject>
#include <QMap>
#include <cstring>
#include <sodium.h>

static const char kRootInfo[] = "messenger-dr-root-v1";
static const int kMaxSkip = 200;

static QByteArray hmacSha256(const QByteArray& key, const QByteArray& data) {
    unsigned char out[crypto_auth_hmacsha256_BYTES];
    crypto_auth_hmacsha256_state st;
    crypto_auth_hmacsha256_init(&st, reinterpret_cast<const unsigned char*>(key.constData()),
                                static_cast<size_t>(key.size()));
    crypto_auth_hmacsha256_update(&st, reinterpret_cast<const unsigned char*>(data.constData()),
                                  static_cast<unsigned long long>(data.size()));
    crypto_auth_hmacsha256_final(&st, out);
    return QByteArray(reinterpret_cast<const char*>(out), crypto_auth_hmacsha256_BYTES);
}

static QByteArray hkdf(const QByteArray& ikm, QByteArray salt, const QByteArray& info, int n) {
    if (salt.isEmpty()) salt = QByteArray(32, '\0');
    const QByteArray prk = hmacSha256(salt, ikm);
    QByteArray t;
    QByteArray out;
    int i = 1;
    while (out.size() < n) {
        QByteArray block = t + info + QByteArray(1, static_cast<char>(i));
        t = hmacSha256(prk, block);
        out += t;
        i++;
    }
    return out.left(n);
}

static bool kdfRk(const QByteArray& rk, const QByteArray& dh, QByteArray& newRk, QByteArray& ck) {
    QByteArray salt = rk.size() == 32 ? rk : QByteArray(32, '\0');
    const QByteArray out = hkdf(dh, salt, QByteArray(kRootInfo), 64);
    if (out.size() != 64) return false;
    newRk = out.left(32);
    ck = out.mid(32, 32);
    return true;
}

static void kdfCk(const QByteArray& ck, QByteArray& next, QByteArray& mk) {
    mk = hmacSha256(ck, QByteArray(1, '\x01'));
    next = hmacSha256(ck, QByteArray(1, '\x02'));
}

static QByteArray x25519(const QByteArray& sk, const QByteArray& pk) {
    if (sk.size() != 32 || pk.size() != 32) return {};
    QByteArray q(32, '\0');
    if (crypto_scalarmult(reinterpret_cast<unsigned char*>(q.data()),
                          reinterpret_cast<const unsigned char*>(sk.constData()),
                          reinterpret_cast<const unsigned char*>(pk.constData())) != 0) {
        return {};
    }
    return q;
}

static bool newKp(QByteArray& pk, QByteArray& sk) {
    pk = QByteArray(crypto_box_PUBLICKEYBYTES, '\0');
    sk = QByteArray(crypto_box_SECRETKEYBYTES, '\0');
    return crypto_box_keypair(reinterpret_cast<unsigned char*>(pk.data()),
                              reinterpret_cast<unsigned char*>(sk.data())) == 0;
}

static QString skipId(const QByteArray& dh, quint32 n) {
    return QString::fromLatin1(dh.toHex()) + QLatin1Char(':') + QString::number(n);
}

static void putBe32(QByteArray& b, int off, quint32 v) {
    b[off] = char((v >> 24) & 0xff);
    b[off + 1] = char((v >> 16) & 0xff);
    b[off + 2] = char((v >> 8) & 0xff);
    b[off + 3] = char(v & 0xff);
}

static quint32 getBe32(const QByteArray& b, int off) {
    return (quint32(uchar(b[off])) << 24) | (quint32(uchar(b[off + 1])) << 16) |
           (quint32(uchar(b[off + 2])) << 8) | quint32(uchar(b[off + 3]));
}

bool DoubleRatchet::initAlice(const QByteArray& theirIk, State& out) {
    if (theirIk.size() != 32) return false;
    QByteArray pk, sk;
    if (!newKp(pk, sk)) return false;
    const QByteArray dh = x25519(sk, theirIk);
    if (dh.isEmpty()) return false;
    QByteArray rk, cks;
    if (!kdfRk(QByteArray(32, '\0'), dh, rk, cks)) return false;
    out = State{};
    out.dhsSk = sk;
    out.dhsPk = pk;
    out.dhr = theirIk;
    out.rk = rk;
    out.cks = cks;
    return true;
}

bool DoubleRatchet::initBob(const QByteArray& myIkPk, const QByteArray& myIkSk, State& out) {
    if (myIkPk.size() != 32 || myIkSk.size() != 32) return false;
    out = State{};
    out.dhsSk = myIkSk;
    out.dhsPk = myIkPk;
    out.rk = QByteArray(32, '\0');
    return true;
}

static bool skipMessageKeys(DoubleRatchet::State& st, quint32 until) {
    if (st.ckr.size() != 32 || st.dhr.size() != 32) return until == 0;
    if (until <= st.nr) return true;
    if (until - st.nr > quint32(kMaxSkip)) return false;
    while (st.nr < until) {
        QByteArray next, mk;
        kdfCk(st.ckr, next, mk);
        st.ckr = next;
        st.skipped.insert(skipId(st.dhr, st.nr), mk);
        st.nr++;
    }
    return true;
}

static bool dhRatchet(DoubleRatchet::State& st, const QByteArray& theirDh) {
    st.pn = st.ns;
    st.ns = 0;
    st.nr = 0;
    st.dhr = theirDh;
    QByteArray dh = x25519(st.dhsSk, st.dhr);
    if (dh.isEmpty()) return false;
    QByteArray rk, ckr;
    if (!kdfRk(st.rk, dh, rk, ckr)) return false;
    st.rk = rk;
    st.ckr = ckr;
    QByteArray pk, sk;
    if (!newKp(pk, sk)) return false;
    st.dhsPk = pk;
    st.dhsSk = sk;
    QByteArray dh2 = x25519(st.dhsSk, st.dhr);
    if (dh2.isEmpty()) return false;
    QByteArray cks;
    if (!kdfRk(st.rk, dh2, rk, cks)) return false;
    st.rk = rk;
    st.cks = cks;
    return true;
}

QByteArray DoubleRatchet::encrypt(State& st, const QByteArray& plain) {
    if (st.cks.size() != 32 || st.dhsPk.size() != 32) return {};
    QByteArray next, mk;
    kdfCk(st.cks, next, mk);
    st.cks = next;
    const quint32 n = st.ns;
    st.ns++;
    QByteArray nonce(24, '\0');
    randombytes_buf(nonce.data(), 24);
    QByteArray ct(plain.size() + crypto_secretbox_MACBYTES, '\0');
    if (crypto_secretbox_easy(reinterpret_cast<unsigned char*>(ct.data()),
                              reinterpret_cast<const unsigned char*>(plain.constData()),
                              static_cast<unsigned long long>(plain.size()),
                              reinterpret_cast<const unsigned char*>(nonce.constData()),
                              reinterpret_cast<const unsigned char*>(mk.constData())) != 0) {
        return {};
    }
    QByteArray out(40 + 24 + ct.size(), '\0');
    memcpy(out.data(), st.dhsPk.constData(), 32);
    putBe32(out, 32, n);
    putBe32(out, 36, st.pn);
    memcpy(out.data() + 40, nonce.constData(), 24);
    memcpy(out.data() + 64, ct.constData(), ct.size());
    return out;
}

QByteArray DoubleRatchet::decrypt(State& st, const QByteArray& payload) {
    if (payload.size() < 40 + 24 + crypto_secretbox_MACBYTES) return {};
    const QByteArray dh = payload.left(32);
    const quint32 n = getBe32(payload, 32);
    const quint32 pn = getBe32(payload, 36);
    const QByteArray nonce = payload.mid(40, 24);
    const QByteArray ct = payload.mid(64);
    const QString id = skipId(dh, n);
    if (st.skipped.contains(id)) {
        const QByteArray mk = st.skipped.take(id);
        QByteArray plain(ct.size() - crypto_secretbox_MACBYTES, '\0');
        if (crypto_secretbox_open_easy(reinterpret_cast<unsigned char*>(plain.data()),
                                       reinterpret_cast<const unsigned char*>(ct.constData()),
                                       static_cast<unsigned long long>(ct.size()),
                                       reinterpret_cast<const unsigned char*>(nonce.constData()),
                                       reinterpret_cast<const unsigned char*>(mk.constData())) != 0) {
            return {};
        }
        return plain;
    }
    if (st.dhr.size() != 32 || st.dhr != dh) {
        if (!skipMessageKeys(st, pn) && st.ckr.size() == 32) return {};
        if (!dhRatchet(st, dh)) return {};
    }
    if (!skipMessageKeys(st, n)) return {};
    QByteArray next, mk;
    kdfCk(st.ckr, next, mk);
    st.ckr = next;
    st.nr++;
    QByteArray plain(ct.size() - crypto_secretbox_MACBYTES, '\0');
    if (crypto_secretbox_open_easy(reinterpret_cast<unsigned char*>(plain.data()),
                                   reinterpret_cast<const unsigned char*>(ct.constData()),
                                   static_cast<unsigned long long>(ct.size()),
                                   reinterpret_cast<const unsigned char*>(nonce.constData()),
                                   reinterpret_cast<const unsigned char*>(mk.constData())) != 0) {
        return {};
    }
    return plain;
}

QString DoubleRatchet::State::toJson() const {
    QJsonObject o;
    o.insert(QStringLiteral("dhs_sk"), QString::fromLatin1(dhsSk.toHex()));
    o.insert(QStringLiteral("dhs_pk"), QString::fromLatin1(dhsPk.toHex()));
    o.insert(QStringLiteral("dhr"), QString::fromLatin1(dhr.toHex()));
    o.insert(QStringLiteral("rk"), QString::fromLatin1(rk.toHex()));
    o.insert(QStringLiteral("cks"), QString::fromLatin1(cks.toHex()));
    o.insert(QStringLiteral("ckr"), QString::fromLatin1(ckr.toHex()));
    o.insert(QStringLiteral("ns"), int(ns));
    o.insert(QStringLiteral("nr"), int(nr));
    o.insert(QStringLiteral("pn"), int(pn));
    o.insert(QStringLiteral("seq"), qint64(seq));
    QJsonObject sk;
    for (auto it = skipped.constBegin(); it != skipped.constEnd(); ++it)
        sk.insert(it.key(), QString::fromLatin1(it.value().toHex()));
    o.insert(QStringLiteral("skipped"), sk);
    return QString::fromUtf8(QJsonDocument(o).toJson(QJsonDocument::Compact));
}

bool DoubleRatchet::State::fromJson(const QString& json, State& out) {
    const QJsonObject o = QJsonDocument::fromJson(json.toUtf8()).object();
    if (o.isEmpty()) return false;
    out = State{};
    out.dhsSk = QByteArray::fromHex(o.value(QStringLiteral("dhs_sk")).toString().toLatin1());
    out.dhsPk = QByteArray::fromHex(o.value(QStringLiteral("dhs_pk")).toString().toLatin1());
    out.dhr = QByteArray::fromHex(o.value(QStringLiteral("dhr")).toString().toLatin1());
    out.rk = QByteArray::fromHex(o.value(QStringLiteral("rk")).toString().toLatin1());
    out.cks = QByteArray::fromHex(o.value(QStringLiteral("cks")).toString().toLatin1());
    out.ckr = QByteArray::fromHex(o.value(QStringLiteral("ckr")).toString().toLatin1());
    out.ns = quint32(o.value(QStringLiteral("ns")).toInt());
    out.nr = quint32(o.value(QStringLiteral("nr")).toInt());
    out.pn = quint32(o.value(QStringLiteral("pn")).toInt());
    out.seq = quint64(o.value(QStringLiteral("seq")).toVariant().toULongLong());
    const QJsonObject sk = o.value(QStringLiteral("skipped")).toObject();
    for (auto it = sk.begin(); it != sk.end(); ++it)
        out.skipped.insert(it.key(), QByteArray::fromHex(it.value().toString().toLatin1()));
    return out.dhsSk.size() == 32 && out.dhsPk.size() == 32;
}

static quint64 drSessionRank(const QString& json) {
    const QJsonObject o = QJsonDocument::fromJson(json.toUtf8()).object();
    const quint64 seq = quint64(o.value(QStringLiteral("seq")).toVariant().toULongLong());
    const quint64 ns = quint64(o.value(QStringLiteral("ns")).toInt());
    const quint64 nr = quint64(o.value(QStringLiteral("nr")).toInt());
    return seq * 1000000ULL + ns + nr;
}

QString DoubleRatchet::State::preferJson(const QString& a, const QString& b) {
    if (a.isEmpty()) return b;
    if (b.isEmpty()) return a;
    return drSessionRank(b) > drSessionRank(a) ? b : a;
}

QMap<QString, QString> DoubleRatchet::State::mergeMaps(const QMap<QString, QString>& local,
                                                       const QMap<QString, QString>& remote) {
    QMap<QString, QString> out = local;
    for (auto it = remote.constBegin(); it != remote.constEnd(); ++it)
        out.insert(it.key(), preferJson(out.value(it.key()), it.value()));
    return out;
}

File: application_android/app/src/main/java/com/messenger/app/data/encryption/DoubleRatchet.kt
package com.messenger.app.data.encryption

import org.json.JSONObject
import java.io.ByteArrayOutputStream
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Direct-chat Double Ratchet (encryption_version=3). Matches back-end/e2ee/ratchet.go.
 * Wire: dh_pk[32] || n_be_u32 || pn_be_u32 || nonce[24] || secretbox(ct).
 */
object DoubleRatchet {
    private const val ROOT_INFO = "messenger-dr-root-v1"
    private const val MAX_SKIP = 200

    class State {
        var dhsSk: ByteArray = ByteArray(0)
        var dhsPk: ByteArray = ByteArray(0)
        var dhr: ByteArray = ByteArray(0)
        var rk: ByteArray = ByteArray(32)
        var cks: ByteArray = ByteArray(0)
        var ckr: ByteArray = ByteArray(0)
        var ns: Int = 0
        var nr: Int = 0
        var pn: Int = 0
        /** Bumped on every local persist; vault merge keeps the higher seq. */
        var seq: Long = 0
        val skipped = HashMap<String, ByteArray>()

        fun toJson(): String {
            val o = JSONObject()
            o.put("dhs_sk", E2ECrypto.toHex(dhsSk))
            o.put("dhs_pk", E2ECrypto.toHex(dhsPk))
            o.put("dhr", if (dhr.size == 32) E2ECrypto.toHex(dhr) else "")
            o.put("rk", E2ECrypto.toHex(rk))
            o.put("cks", if (cks.size == 32) E2ECrypto.toHex(cks) else "")
            o.put("ckr", if (ckr.size == 32) E2ECrypto.toHex(ckr) else "")
            o.put("ns", ns)
            o.put("nr", nr)
            o.put("pn", pn)
            o.put("seq", seq)
            val sk = JSONObject()
            skipped.forEach { (k, v) -> sk.put(k, E2ECrypto.toHex(v)) }
            o.put("skipped", sk)
            return o.toString()
        }

        companion object {
            fun fromJson(s: String): State? {
                return try {
                    val o = JSONObject(s)
                    val st = State()
                    st.dhsSk = E2ECrypto.fromHex(o.optString("dhs_sk")) ?: return null
                    st.dhsPk = E2ECrypto.fromHex(o.optString("dhs_pk")) ?: return null
                    val dhr = o.optString("dhr")
                    st.dhr = if (dhr.length == 64) E2ECrypto.fromHex(dhr) ?: ByteArray(0) else ByteArray(0)
                    st.rk = E2ECrypto.fromHex(o.optString("rk")) ?: ByteArray(32)
                    val cks = o.optString("cks")
                    st.cks = if (cks.length == 64) E2ECrypto.fromHex(cks) ?: ByteArray(0) else ByteArray(0)
                    val ckr = o.optString("ckr")
                    st.ckr = if (ckr.length == 64) E2ECrypto.fromHex(ckr) ?: ByteArray(0) else ByteArray(0)
                    st.ns = o.optInt("ns")
                    st.nr = o.optInt("nr")
                    st.pn = o.optInt("pn")
                    st.seq = o.optLong("seq")
                    val sk = o.optJSONObject("skipped")
                    if (sk != null) {
                        val keys = sk.keys()
                        while (keys.hasNext()) {
                            val k = keys.next()
                            val v = E2ECrypto.fromHex(sk.optString(k)) ?: continue
                            st.skipped[k] = v
                        }
                    }
                    st
                } catch (_: Exception) {
                    null
                }
            }
        }
    }

    fun initAlice(theirIk: ByteArray): State? {
        val kp = E2ECrypto.generateKeyPair() ?: return null
        val sk = E2ECrypto.fromHex(kp.privateHex) ?: return null
        val pk = E2ECrypto.fromHex(kp.publicHex) ?: return null
        val dh = E2ECrypto.x25519(sk, theirIk) ?: return null
        val (rk, cks) = kdfRk(ByteArray(32), dh) ?: return null
        val st = State()
        st.dhsSk = sk
        st.dhsPk = pk
        st.dhr = theirIk.copyOf()
        st.rk = rk
        st.cks = cks
        return st
    }

    fun initBob(myIkPk: ByteArray, myIkSk: ByteArray): State {
        val st = State()
        st.dhsSk = myIkSk.copyOf()
        st.dhsPk = myIkPk.copyOf()
        st.rk = ByteArray(32)
        return st
    }

    fun encrypt(st: State, plain: ByteArray): ByteArray? {
        if (st.cks.size != 32 || st.dhsPk.size != 32) return null
        val (next, mk) = kdfCk(st.cks)
        st.cks = next
        val n = st.ns
        st.ns++
        val nonce = ByteArray(24)
        java.security.SecureRandom().nextBytes(nonce)
        val ct = E2ECrypto.secretBoxSeal(plain, nonce, mk) ?: return null
        val out = ByteArray(40 + 24 + ct.size)
        System.arraycopy(st.dhsPk, 0, out, 0, 32)
        putBe32(out, 32, n)
        putBe32(out, 36, st.pn)
        System.arraycopy(nonce, 0, out, 40, 24)
        System.arraycopy(ct, 0, out, 64, ct.size)
        return out
    }

    fun decrypt(st: State, payload: ByteArray): ByteArray? {
        if (payload.size < 40 + 24 + 16) return null
        val dh = payload.copyOfRange(0, 32)
        val n = getBe32(payload, 32)
        val pn = getBe32(payload, 36)
        val nonce = payload.copyOfRange(40, 64)
        val ct = payload.copyOfRange(64, payload.size)
        val id = skipId(dh, n)
        st.skipped.remove(id)?.let { mk ->
            return E2ECrypto.secretBoxOpen(ct, nonce, mk)
        }
        if (st.dhr.size != 32 || !st.dhr.contentEquals(dh)) {
            if (!skipMessageKeys(st, pn) && st.ckr.size == 32) return null
            if (!dhRatchet(st, dh)) return null
        }
        if (!skipMessageKeys(st, n)) return null
        val (next, mk) = kdfCk(st.ckr)
        st.ckr = next
        st.nr++
        return E2ECrypto.secretBoxOpen(ct, nonce, mk)
    }

    private fun skipMessageKeys(st: State, until: Int): Boolean {
        if (st.ckr.size != 32 || st.dhr.size != 32) return until == 0
        if (until <= st.nr) return true
        if (until - st.nr > MAX_SKIP) return false
        while (st.nr < until) {
            val (next, mk) = kdfCk(st.ckr)
            st.ckr = next
            st.skipped[skipId(st.dhr, st.nr)] = mk
            st.nr++
        }
        return true
    }

    private fun dhRatchet(st: State, theirDh: ByteArray): Boolean {
        st.pn = st.ns
        st.ns = 0
        st.nr = 0
        st.dhr = theirDh.copyOf()
        val dh = E2ECrypto.x25519(st.dhsSk, st.dhr) ?: return false
        val (rk, ckr) = kdfRk(st.rk, dh) ?: return false
        st.rk = rk
        st.ckr = ckr
        val kp = E2ECrypto.generateKeyPair() ?: return false
        st.dhsSk = E2ECrypto.fromHex(kp.privateHex) ?: return false
        st.dhsPk = E2ECrypto.fromHex(kp.publicHex) ?: return false
        val dh2 = E2ECrypto.x25519(st.dhsSk, st.dhr) ?: return false
        val (rk2, cks) = kdfRk(st.rk, dh2) ?: return false
        st.rk = rk2
        st.cks = cks
        return true
    }

    private fun hmac(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data)
    }

    private fun kdfRk(rk: ByteArray, dh: ByteArray): Pair<ByteArray, ByteArray>? {
        val salt = if (rk.size == 32) rk else ByteArray(32)
        val out = hkdf(dh, salt, ROOT_INFO.toByteArray(), 64)
        return out.copyOfRange(0, 32) to out.copyOfRange(32, 64)
    }

    private fun kdfCk(ck: ByteArray): Pair<ByteArray, ByteArray> {
        val mk = hmac(ck, byteArrayOf(0x01))
        val next = hmac(ck, byteArrayOf(0x02))
        return next to mk
    }

    private fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, n: Int): ByteArray {
        val prk = hmac(salt, ikm)
        var t = ByteArray(0)
        val buf = ByteArrayOutputStream()
        var i = 1
        while (buf.size() < n) {
            t = hmac(prk, t + info + byteArrayOf(i.toByte()))
            buf.write(t)
            i++
        }
        return buf.toByteArray().copyOf(n)
    }

    private fun skipId(dh: ByteArray, n: Int) = E2ECrypto.toHex(dh) + ":" + n

    private fun putBe32(b: ByteArray, off: Int, v: Int) {
        b[off] = ((v ushr 24) and 0xff).toByte()
        b[off + 1] = ((v ushr 16) and 0xff).toByte()
        b[off + 2] = ((v ushr 8) and 0xff).toByte()
        b[off + 3] = (v and 0xff).toByte()
    }

    private fun getBe32(b: ByteArray, off: Int): Int {
        return ((b[off].toInt() and 0xff) shl 24) or
            ((b[off + 1].toInt() and 0xff) shl 16) or
            ((b[off + 2].toInt() and 0xff) shl 8) or
            (b[off + 3].toInt() and 0xff)
    }

    /** Rank for vault merge: seq first, then send+receive counters. Ties keep [a]. */
    fun preferSessionJson(a: String, b: String): String {
        if (a.isBlank()) return b
        if (b.isBlank()) return a
        return if (sessionRank(b) > sessionRank(a)) b else a
    }

    fun mergeSessionMaps(local: Map<String, String>, remote: Map<String, String>): Map<String, String> {
        val out = local.toMutableMap()
        for ((k, v) in remote) {
            out[k] = preferSessionJson(out[k].orEmpty(), v)
        }
        return out
    }

    private fun sessionRank(json: String): Long {
        return try {
            val o = JSONObject(json)
            o.optLong("seq") * 1_000_000L + o.optLong("ns") + o.optLong("nr")
        } catch (_: Exception) {
            0L
        }
    }
}
